dev/OAuth/AuthRequestAsyncOperation.cpp

```diff
@@ -0,0 +1,574 @@
+            throw winrt::hresult_error(HRESULT_FROM_WIN32(err), msg);
+        }
+
+        m_overlapped.hEvent = ::CreateEventW(nullptr, true, false, nullptr);
+        if (!m_overlapped.hEvent)
+        {
+            throw winrt::hresult_error(HRESULT_FROM_WIN32(::GetLastError()), L"Failed to create an event");
+        }
+
+        m_ptp = ::CreateThreadpoolWait(async_callback, this, nullptr);
+        connect_to_new_client();
+    }
+    catch (...)
+    {
+        // Throwing in a constructor will cause the destructor not to run...
+        destroy();
+        throw;
+    }
+}
+
+AuthRequestAsyncOperation::~AuthRequestAsyncOperation()
+{
+    destroy();
+}
+
+void AuthRequestAsyncOperation::destroy()
+{
+    {
+        // Expects lock to be held and is required since we haven't ensured all callbacks have completed
+        std::unique_lock guard{ m_mutex };
+        close_pipe();
+    }
+
+    // Note that we don't hold the lock here for two reasons. The big reason is that 'WaitForThreadpoolWaitCallbacks may
+    // wait on a callback trying to acquire the lock. The second reason - and the reason we get away with this - is that
+    // this code path only gets called on destruction, meaning nothing except callbacks (which we wait for) will access
+    // or modify object state
+    if (m_ptp)
+    {
+        if (!::SetThreadpoolWaitEx(m_ptp, nullptr, nullptr, nullptr))
+        {
+            // False here means that there's a callback in progress. This would realistically only happen if there was
+            // a race between the client calling 'Cancel' and someone connecting to the pipe
+            ::WaitForThreadpoolWaitCallbacks(m_ptp, true);
+        }
+
+        ::CloseThreadpoolWait(m_ptp);
+        m_ptp = nullptr;
+    }
+
+    if (m_overlapped.hEvent)
+    {
+        ::CloseHandle(m_overlapped.hEvent);
+        m_overlapped.hEvent = nullptr;
+    }
+}
+
+void AuthRequestAsyncOperation::close_pipe()
+{
+    auto lastState = std::exchange(m_state, request_state::closed);
+    if (lastState == request_state::closed)
+    {
+        return;
+    }
+
+    if (m_pipe != INVALID_HANDLE_VALUE)
+    {
+        ::CancelIoEx(m_pipe, &m_overlapped);
+        ::CloseHandle(m_pipe);
+        m_pipe = INVALID_HANDLE_VALUE;
+    }
+}
+
+winrt::hresult AuthRequestAsyncOperation::ErrorCode()
+{
+    std::shared_lock guard{ m_mutex };
+    return m_error;
+}
+
+uint32_t AuthRequestAsyncOperation::Id()
+{
+    return 1; // NOTE: This is copying the C++/WinRT implementation
+}
+
+winrt::Windows::Foundation::AsyncStatus AuthRequestAsyncOperation::Status()
+{
+    std::shared_lock guard{ m_mutex };
+    return m_status;
+}
+
+void AuthRequestAsyncOperation::Cancel()
+{
+    winrt::make_self<factory_implementation::OAuth2Manager>()->cancel(this);
+}
+
+void AuthRequestAsyncOperation::Close()
+{
+    // TODO? C++/WinRT does a noop here
```

Comment:
What is this comment? Do we need the Close() function?

Reply:
The current implementation does nothing (a no-op, or "no operation") in this context.
This method is part of the IAsyncInfo interface.